File: Dispatcher/MessageDispatcher.m
//
//  MessageDispatcher.m
//  Created by Gal Blank on 9/21/15.
//  Copyright © 2015 Gal Blank. All rights reserved.
//

#import "MessageDispatcher.h"
#import <pos-Swift.h>
#import "POS-Bridging-Header.h"
#import "RegexKitLite.h"

@implementation MessageDispatcher

static MessageDispatcher *sharedDispatcherInstance = nil;


+ (id)sharedInstance
{
    static dispatch_once_t once;
    static id sharedInstance;
    dispatch_once(&once, ^{
        sharedInstance = [[self alloc] init];
    });
    return sharedInstance;
}


+ (id)allocWithZone:(NSZone *)zone {
    @synchronized(self) {
        if (sharedDispatcherInstance == nil) {
            sharedDispatcherInstance = [super allocWithZone:zone];
            // assignment and return on first allocation
            return sharedDispatcherInstance;
        }
    }
    // on subsequent allocation attempts return nil
    return nil;
}

- (id)copyWithZone:(NSZone *)zone
{
    return self;
}

-(id)init
{
    if (self = [super init]) {
        if(messageBus == nil){
            messageBus = [[NSMutableArray alloc] init];
        }
        
        if(dispatchedMessages == nil){
            dispatchedMessages = [[NSMutableArray alloc] init];
        }
        
        
             [NSTimer scheduledTimerWithTimeInterval:CLEANUP_TIMER target:self selector:@selector(clearDispastchedMessages) userInfo:nil repeats:YES];
        
       
    }
    return self;
}

-(void)addMessageToBus:(Message*)newmessage
{

    if(newmessage.ttl == DEFAULT_TTL){
        [messageBus addObject:newmessage];
        if(dispsatchTimer == nil){
            [self startDispatching];
        }
    }
    else{
        [self dispatchMessage:newmessage];
    }
}


-(void)clearDispastchedMessages
{
    for (Message *msg in dispatchedMessages) {
        [messageBus removeObject:msg];
    }
    [dispatchedMessages removeAllObjects];
}

-(void)dispatchThisMessage:(NSTimer*)timer
{
    Message* message = [timer.userInfo objectForKey:@"message"];
    if(message){
        [self dispatchMessage:message];
    }
}

-(void)startDispatching
{
    dispsatchTimer = [NSTimer scheduledTimerWithTimeInterval:DEFAULT_TTL target:self selector:@selector(leave) userInfo:nil repeats:YES];
}

-(void)stopDispathing
{
    if(dispsatchTimer){
        [dispsatchTimer invalidate];
        dispsatchTimer = nil;
    }
}

-(void)leave
{
    NSArray * goingAwayBus = [NSArray arrayWithArray:messageBus];
    for(Message *msg in goingAwayBus){
        [self dispatchMessage:msg];
    }
}


<<<<<<< HEAD
-(messageType)messageNameTomessageType:(NSString*)messageName
{
    if([messageName caseInsensitiveCompare:@"TokenForTransaction"] == NSOrderedSame){
        return messageTypeTokenForTransaction;
    }
    else if([messageName caseInsensitiveCompare:@"IngenicoMessage"] == NSOrderedSame){
        return messageTypeIngenicoMessage;
    }
    else if([messageName caseInsensitiveCompare:@"OnHoldOrderMessage"] == NSOrderedSame){
        return messageTypeIngenicoMessage;
    }
    
    return -1;
}

-(NSString*)messageTypeToString:(messageType)Type
{
    NSString *retMessage = @"";
    switch (Type) {
        case messageTypeMESSAGETYPE_GET_CONFIG:
            retMessage = @"MESSAGETYPE_GET_CONFIG";
            break;
        case messageTypeTokenForTransactionRequest:
            retMessage = @"TokenForTransactionRequest";
            break;
        case messageTypeTokenForTransaction:
            retMessage = @"TokenForTransaction";
            break;
        case messageTypeIngenicoMessage:
            retMessage = @"IngenicoMessage";
            break;
        case messageTypeUserInitializeHardware:
            return @"messageTypeUserInitializeHardware";
            break;
        case messageTypeStartScanners:
            return @"messageTypeStartScanners";
            break;
        case messageTypeStopScanners:
            return @"messageTypeStopScanners";
            break;
        case messageTypeProductScanned:
            return @"messageTypeProductScanned";
            break;
        case messageTypeOnHoldOrdersBatchSync:
            return @"messageTypeOnHoldOrdersBatchedSync";
            break;
        case messageTypeOnHoldOrderSync:
            return @"messageTypeOnHoldOrderSync";
        case messageTypeGotoPortal:
            return @"messageTypeGotoPortal";
            break;
        case messageTypeComebackFromPortal:
            return @"messageTypeComebackFromPortal";

            break;
        default:
            break;
    }
    
    return retMessage;
}
=======
>>>>>>> a48ee7ba9c2880518082747d8c1aacba9abdbe8e

-(void)dispatchMessage:(Message*)message
{
    NSMutableDictionary * messageDic = [[NSMutableDictionary alloc] init];
    
    if([[message routeFromRoutingKey] caseInsensitiveCompare:@"api"] == NSOrderedSame){
        [MessageApiConverter.sharedInstance messageTypeToApiCall:message];
    }
    
    [messageDic setObject:message forKey:@"message"];
    [[NSNotificationCenter defaultCenter] postNotificationName:message.routingKey object:nil userInfo:messageDic];
    [dispatchedMessages addObject:message];
}


-(void)routeMessageToServerWithType:(Message*)message
{
    if(message.params == nil){
        message.params = [[NSMutableDictionary alloc] init];
    }
    
    NSString * sectoken = [[NSUserDefaults standardUserDefaults] objectForKey:@"securitytoken"];
    
    if(sectoken && sectoken.length > 0){
        [message.params setObject:sectoken forKey:@"securitytoken"];
    }
}




-(BOOL)canSendMessage:(Message*)message
{
    return YES;
}


@end

File: Dispatcher/MessageDispatcher.swift
//
//  MessageDispatcher.swift
//  POS
//
//  Created by Gal Blank on 1/15/16.
//

import UIKit

class MessageDispatcher: NSObject {
    
    fileprivate lazy var dispatchOnce: () -> Void = {
        DispatchQueue.main.async(execute: { () -> Void in
            if self.dispsatchTimer == nil {
                self.startDispatching()
            }
        })
        return {}
    }()
    
    @objc static let sharedDispacherInstance = MessageDispatcher()
    
    var dispsatchTimer:Timer?
    var messageBus         = [Message]()
    var dispatchedMessages = [Message]()
    
    struct Static {
        static var token: Int = 0
    }
    
    @objc func consumeMessage(_ notif:Foundation.Notification) {
        let msg:Message = notif.userInfo!["message"] as! Message
        switch(msg.routingKey){
        case "msg.selfdestruct":
            if let index = messageBus.index(of: msg) {
                if(index >= 0){
                    messageBus.remove(at: index)
                }
            }
            break
        default:
            break
        }
    }
    
    @objc func addMessageToBus(_ newmessage: Message) {
        
        DispatchQueue.main.async {
            if(newmessage.shouldselfdestruct == false && newmessage.routingKey.caseInsensitiveCompare("msg.selfdestruct") == ComparisonResult.orderedSame) {
                if let index = self.messageBus.index(of: newmessage) {
                    if(index >= 0 ) {
                        self.messageBus.remove(at: index)
                    }
                }
            }
            self.messageBus.append(newmessage)
            self.dispatchOnce()
        }
    }
    
    func clearDispastchedMessages() {
        
        DispatchQueue.main.async {
            for msg:Message in self.dispatchedMessages {
                if let index = self.messageBus.index(of: msg) {
                    if(index >= 0) {
                        self.messageBus.remove(at: index)
                    }
                }
            }
            self.dispatchedMessages.removeAll()
        }
    }
    
    
    func startDispatching() {
        NotificationCenter.default.addObserver(self,
                                               selector: #selector(MessageDispatcher.consumeMessage(_:)),
                                               name: NSNotification.Name(rawValue: "msg.selfdestruct"), object: nil)
        
        dispsatchTimer = Timer.scheduledTimer(timeInterval: 0.1, target: self, selector: #selector(MessageDispatcher.leave), userInfo: nil, repeats: true)
    }
    
    func stopDispathing() {
        if dispsatchTimer != nil {
            dispsatchTimer!.invalidate()
            dispsatchTimer = nil
        }
    }
    
    @objc func leave() {
        DispatchQueue.main.async {
            let goingAwayBus = NSArray(array: self.messageBus) as! [Message]
            
            for msg: Message in goingAwayBus {
                if(msg.shouldselfdestruct == false) {
                    self.dispatchMessage(msg)
                    msg.shouldselfdestruct = true
                    
                    if let index = self.messageBus.index(of: msg) {
                        if(index != NSNotFound) {
                            self.messageBus.remove(at: index)
                        }
                    }
                }
            }
        }
    }
    
    func dispatchMessage(_ message: Message) {
        
        var messageDic = [AnyHashable: Any]()
        
        if message.routeFromRoutingKey().caseInsensitiveCompare("api") == ComparisonResult.orderedSame {
            MessageApiConverter.sharedInstance.messageTypeToApiCall(message)
        }
        
        messageDic["message"] = message
        NotificationCenter.default.post(name: Notification.Name(rawValue: message.routingKey), object: nil, userInfo: messageDic)
    }
    
    func routeMessageToServerWithType(_ message: Message) {
        
        if message.params == nil {
            message.params? = [AnyHashable: Any]() as AnyObject
        }
        
        if let sectoken = UserDefaults.standard.object(forKey: "securitytoken") as? String {
            if sectoken.lengthOfBytes(using: String.Encoding.utf8) > 0 {
                message.params?.set(sectoken, forKey: "securitytoken")
            }
        }
    }
    
    func canSendMessage(_ message: Message) -> Bool {
        return true
    }
}
